### ISSUE TITLE
large file is going to blow memory

### ISSUE THREAD

**Issue by philister** [carrierwaveuploader/carrierwave#1780]
I believe the commit 927479c84e968ed03546c7fdd16107beeb771a22 (PR #1517 and issue #1338) leads to a memory-problem when uploading large files: 
I guess `file.read` reads the complete file-content into memory. Fog can not "stream" the file from disk, using the multipart-feature correctly (when set `fog_attributes[:multipart_chunk_size]`)
At least, when I revert this commit, my large files (more GB than available mem) are uploaded. Current master returns a "NoMemoryError: failed to allocate memory".

**Comment by thomasfedb** [comment]
Hi @philister, can you provide an example of an uploader that exhibits this issue?

**Comment by philister** [comment]
Gemfile.lock (excerpt)
``` 
remote: git@github.com:carrierwaveuploader/carrierwave.git
revision: d6e7a4480ae595ad28f4cd19c5b1c6a7918b015f

rails (4.2.5)
fog (1.36.0)
```
animal.rb
```
class Animal < ActiveRecord::Base
  mount_uploader :picture
end
```

rails c
```
a = Animal.new; a.picture = File.open('10GB.zip') # more than RAM
=> #<File:/home/philipp/testrails/10GB.zip>
>> a.save
   (0.1ms)  begin transaction
  SQL (0.8ms)  INSERT INTO "animals" ("picture", "created_at", "updated_at") VALUES (?, ?, ?)  [["picture", "10GB.zip"], ["created_at", "2015-11-25 10:46:04.552696"], ["updated_at", "2015-11-25 10:46:04.552696"]]
   (0.2ms)  rollback transaction
NoMemoryError: failed to allocate memory
	from /home/philipp/.rvm/gems/ruby-2.2.3/bundler/gems/carrierwave-d6e7a4480ae5/lib/carrierwave/storage/fog.rb:315:in `read
```

**Comment by thomasfedb** [comment]
@philister Can you provide a full stack trace please?

**Comment by philister** [comment]
Added above.

**Comment by thomasfedb** [comment]
@bensie Are you able to look over this?

**Comment by bensie** [comment]
This definitely seems problematic, though I'm not sure when I'll have time to dig in and take a look.

**Comment by leandro** [comment]
I'm going through a similar problem but on the opposite direction. I'm fetching a large file and I want to be able to (according to @philister example) forward it to the request's response by chunks:

```ruby
tmp_path = Tempfile.create('animal-tmp') do |output|
  while buffer = animal.picture.read(4096)
    output.write(buffer)
  end
  output.path
end

send_file tmp_path, type: animal.picture.content_type, disposition: 'inline'
```

And I noticed, that the `CarrierWave::Uploader::Proxy#read` doesn't allow me to do so:
https://github.com/carrierwaveuploader/carrierwave/blob/master/lib/carrierwave/uploader/proxy.rb#L43

PS.: I'm not using `send_data` because using `send_file` [is apparently faster](http://stackoverflow.com/a/26884350/1170086).

**Comment by leandro** [comment]
Sorry. Never mind. `animal.picture.file` points to a `CarrierWave::Storage::Fog::File`'s instance. So I'll check out their code to see what I can do about it.

**Comment by tmedford** [comment]
Is there any update on this?

**Comment by mshibuya** [comment]
Fixed by #2314.
